File: api/database/models/Category.py
from typing import List, Optional
from sqlmodel import SQLModel, Field, Relationship
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .TagLabel import TagLabel, TagLabelRead
    from .MenuItem import MenuItem, MenuItemRead
    from .Restaurant import Restaurant

class CategoryBase(SQLModel):
    name: str = Field(unique=True, index=True)
    restaurant_id: int = Field(foreign_key="restaurants.id", index=True)  # Link to a specific restaurant

class Category(CategoryBase, table=True):
    __tablename__: str = "categories"

    id: Optional[int] = Field(default=None, primary_key=True, index=True, nullable=False)
    parent_id: Optional[int] = Field(default=None, foreign_key="categories.id", nullable=True)

    parent: Optional["Category"] = Relationship(
        back_populates="children",
        sa_relationship_kwargs=dict(remote_side="Category.id")
    )

    children: List["Category"] = Relationship(back_populates="parent")
    menu_items: List['MenuItem'] = Relationship(back_populates="category")
    tag_labels: List['TagLabel'] = Relationship(back_populates="category")

class CategoryLite(CategoryBase):
    id: int
    parent: Optional['CategoryLite'] = None

class CategoryRead(CategoryBase):
    id: int
    parent: Optional['CategoryLite'] = None
    menu_items: List['MenuItemRead'] = []
    tag_labels: List['TagLabelRead'] = []

class CategoryTreeRead(CategoryBase):
    id: int
    parent: Optional['CategoryLite'] = None
    children: List["CategoryRead"] = []
    menu_items: List['MenuItemRead'] = []
    tag_labels: List['TagLabelRead'] = []

    def toCategoryRead(self) -> CategoryRead:
        return CategoryRead(
            name=self.name,
            menu_items=self.menu_items,
            id=self.id,
            parent_id=self.parent_id
        )
    
class CategoryTreeLite(CategoryBase):
    id: int
    children: List["CategoryTreeLite"] = []
    tag_labels: List['TagLabelRead'] = []

class CategoryCreate(CategoryBase):
    restaurant_id: int
    parent_id: int

File: src/app/types/category.ts
import MenuItem from '@/app/types/menuItem';
import { TagLabel } from '@/app/types/tag';

export class CategoryLite {
  constructor(
    public id: number,
    public restaurant_id: number,
    public name: string,
    public parent: CategoryLite | null,
  ) { }

  static fromObject(object: any): CategoryLite {
    const { id, restaurant_id, name, parent } = object;
    return new CategoryLite(id, restaurant_id, name, parent && this.fromObject(parent));
  }
}

export class Category {
  constructor(
    public id: number,
    public restaurant_id: number,
    public name: string,
    public parent: CategoryLite | null,
    public menu_items: MenuItem[] = [],
    public tag_labels: TagLabel[] = []
  ) { }

  static fromObject(object: any): Category {
    const { id, restaurant_id, name, parent, menu_items, tag_labels } = object;
    return new Category(id,
      restaurant_id,
      name,
      parent && CategoryLite.fromObject(parent),
      menu_items,
      tag_labels);
  }

  static toCategoryLite(category: Category) {
    return new CategoryLite(
      category.id,
      category.restaurant_id,
      category.name,
      category.parent
    )
  }
}

export class CategoryTree {
  constructor(
    public id: number,
    public restaurant_id: number,
    public name: string,
    public parent: CategoryLite | null,
    public children: Category[] = [],
    public menu_items: MenuItem[] = [],
    public tag_labels: TagLabel[] = []
  ) { }

  static toCategoryType(categoryTree: CategoryTree): Category {
    return new Category(categoryTree.id, 
      categoryTree.restaurant_id,
      categoryTree.name,
      categoryTree.parent, 
      categoryTree.menu_items, 
      categoryTree.tag_labels);
  }

  static fromObject(object: any): CategoryTree {
    const { id, restaurant_id, name, parent, children, menu_items, tag_labels } = object;
    return new CategoryTree(
      id,
      restaurant_id,
      name,
      parent && CategoryLite.fromObject(parent),
      children ? children.map(Category.fromObject) : [],
      menu_items,
      tag_labels
    );
  }
}

export class CategoryCreate {
  constructor(
    public name: string,
    public restaurant_id: number,
    public parent_id: number
  ) { }
}

export class CategoryTreeLite {
  constructor(
    public id: number,
    public restaurant_id: number,
    public name: string,
    public children: CategoryTreeLite[] = [],
  ) { }

  static fromObject(object: any): CategoryTreeLite {
    const { id, restaurant_id, name, children } = object;
    return new CategoryTreeLite(
      id,
      restaurant_id,
      name,
      children ? children.map(CategoryTreeLite.fromObject) : [],
    );
  }
}

